src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -142,35 +142,72 @@ private BoundExpression CompareWithConstant(BoundExpression input, BoundExpressi
+            var conversionKind = _compilation.Conversions.ClassifyConversionFromType(sourceType, targetType, ref ignoredDiagnostics).Kind;
+            var constantResult = Binder.GetIsOperatorConstantResult(sourceType, targetType, conversionKind, source.ConstantValue, requiredNullTest);
+            return
+                constantResult == ConstantValue.True ? true :
+                constantResult == ConstantValue.False ? false :
+                constantResult == null ? (bool?)null :
+                throw ExceptionUtilities.UnexpectedValue(constantResult);
         }
 
         BoundExpression MakeIsDeclarationPattern(SyntaxNode syntax, BoundExpression loweredInput, BoundExpression loweredTarget, bool requiresNullTest)
         {
             var type = loweredTarget.Type;
             requiresNullTest = requiresNullTest && loweredInput.Type.CanContainNull();
 
+            // If the match is impossible, we simply evaluate the input and yield false.
+            var matchConstantValue = MatchConstantValue(loweredInput, type, false);
+            if (matchConstantValue == false)
+            {
+                return _factory.MakeSequence(loweredInput, _factory.Literal(false));
+            }
+
             // It is possible that the input value is already of the correct type, in which case the pattern
             // is irrefutable, and we can just do the assignment and return true (or perform the null test).
-            if (MatchIsIrrefutable(loweredInput.Type, loweredTarget.Type, requiresNullTest))
+            if (matchConstantValue == true)
             {
-                var convertedInput = _factory.Convert(loweredTarget.Type, loweredInput);
-                var assignment = _factory.AssignmentExpression(loweredTarget, convertedInput);
-                return requiresNullTest
-                    ? _factory.ObjectNotEqual(assignment, _factory.Null(type))
-                    : _factory.MakeSequence(assignment, _factory.Literal(true));
+                requiresNullTest = requiresNullTest && MatchConstantValue(loweredInput, type, true) != true;
+                if (loweredInput.Type.IsNullableType())
+                {
+                    var getValueOrDefault = _factory.SpecialMethod(SpecialMember.System_Nullable_T_GetValueOrDefault).AsMember((NamedTypeSymbol)loweredInput.Type);
+                    if (requiresNullTest)
+                    {
+                        //bool Is<T>(T? input, out T output) where T : struct
+                        //{
+                        //    output = input.GetValueOrDefault();
+                        //    return input.HasValue;
+                        //}
+
+                        var input = _factory.SynthesizedLocal(loweredInput.Type, syntax); // we copy the input to avoid double evaluation
+                        var getHasValue = _factory.SpecialMethod(SpecialMember.System_Nullable_T_get_HasValue).AsMember((NamedTypeSymbol)loweredInput.Type);
```

Comment:
Also see https://github.com/dotnet/roslyn/pull/19413/commits/9e05b042ba33f7ec3de69f46a067ee9491dc47a8 which adds a test that triggers the requested condition.